### PR TITLE
Allow passing static shape to tensor creation helpers

### LINKED ISSUE #61: Allow shape specification in tensor constructor helpers
### Please describe the purpose of filing this issue

Would be nice if we could pass the static shape to the tensor constructors now that it is supported

```python
import pytensor.tensor as pt
pt.vector(shape=(3, 4))
```

For some types that constrain the shape, we could assert they are compatible
```python
pt.col(shape=(5, 1))
pt.col(shape=(1, 5))  # Raise ValueError
```

In addition, would be nice if the more general `at.tensor`, had a default dtype of `floatX`, which is the most common case

```python
pt.tensor("float64", shape=(None, None, 5))  # Fine
pt.tensor(shape=(None, None, 5))
# TypeError: TensorType.__init__() missing 1 required positional argument: 'dtype'
```

These are implemented here: https://github.com/pymc-devs/pytensor/blob/b7a2d30830da8b201ab47879699a5a7553193fdc/pytensor/tensor/type.py#L778-L1118

### PR DESCRIPTION
**Pull request by ricardoV94**
Closes #61

## Major / Breaking Changes
- `pytensor.tensor.tensor` now accepts name as the only positional argument. All other arguments must be provided via keyword

## New features
- Tensor creator helpers now accept static input shapes (e.g., `pt.matrix(shape=(10, None))`)

### CODE REVIEW

**Review by ferrine — CHANGES_REQUESTED** [review]

File: pytensor/tensor/type.py [hunk 1]
```diff
@@ -775,9 +776,17 @@ def values_eq_approx_always_true(a, b):
 )
 
 
-def tensor(*args, **kwargs):
+def tensor(
+    dtype: Optional["DTypeLike"] = None,
```
[the remark this answers is not in the document]
Comment (ricardoV94) [reply pytensor/tensor/type.py:780]: I didn't want to modify the signature here, just introduce a default `dtype`

[the remark this answers is not in the document]
Comment (ricardoV94) [reply pytensor/tensor/type.py:780]: `tensor` is used in 234 places :)

[the remark this answers is not in the document]
Comment (ricardoV94) [reply pytensor/tensor/type.py:780]: tensor is used in 234 places :)

[the remark this answers is not in the document]
Comment (ferrine) [reply pytensor/tensor/type.py:780]: is it possible to do the other way round? adding the same api to all other helpers?

**Review by ferrine — APPROVED** [review]
Happy with the changes!

File: pytensor/tensor/type.py [hunk 2]
```diff
@@ -794,7 +817,11 @@ def tensor(*args, **kwargs):
 ulscalar = TensorType("uint64", ())
 
 
-def scalar(name=None, dtype=None):
+def scalar(
+    name: Optional[str] = None,
+    *,
+    dtype: Optional["DTypeLike"] = None,
+) -> "TensorVariable":
```
Comment (michaelosthege) [inline pytensor/tensor/type.py:824]: Is this the public API `pt.scalar()` ?

Why the breaking change?

File: tests/tensor/test_type.py [hunk 3]
```diff
@@ -326,3 +339,114 @@ def test_deprecated_kwargs():
...
+    assert res.name is None
+
+    res = tensor(dtype="int64", shape=(5, None), name="custom")
+    assert res.type == TensorType(dtype="int64", shape=(5, None))
+    assert res.name == "custom"
+
+    # Test with positional arguments only
+    res = tensor(dtype="int64", shape=(5, None))
+    assert res.type == TensorType(dtype="int64", shape=(5, None))
+    assert res.name is None
```
Comment (michaelosthege) [inline tests/tensor/test_type.py:356]: This is an exact duplicate of the block above

File: pytensor/tensor/type.py (hunk 2 above)
Comment (ricardoV94) [reply pytensor/tensor/type.py:824]: You mean forcing `dtype` to be kwarg only?

**Review by michaelosthege — CHANGES_REQUESTED** [review]
I found one thing in a test

generally I'm worried that the breaking change to require explicit kwargs will cause a lot of things to break downstream.

Requiring `shape` to be an explicit kwarg (but not dtype) should be much smoother, no?

File: tests/tensor/test_type.py (hunk 3 above)
Comment (ricardoV94) [reply tests/tensor/test_type.py:356]: You're right. I missed it from the changes between the first and second commits

**Comment by ferrine** [comment]
Yes, I believe uniform API is much better long term, and since we anyway just made the fork, we better make these changes sooner than later

**Review by michaelosthege — APPROVED** [review]
Okay, I just checked the PyMC codebase and found no places where it would actually break.

**Pull request merged**